src/TTextEdit.cpp

```diff
@@ -414,9 +421,9 @@ void TTextEdit::drawLine(QPainter& painter, int lineNumber, int lineOfScreen) co
  * @param grapheme
  * @param column Used to calculate the width of Tab
  * @param charStyle
- * @return Return the display width of the grapheme
+ * @return Return the display width of the grapheme and the width in pixels need to display it
```

Comment:
Why do these two widths differ?

Reply:
The first is the *number of **normal (width)** spaces* that code in `widechar_wide.h` tells us and will be 1 or 2 in the case of normal printable graphemes; the second is the amount of pixels that the `QPainter` instance tells us it used to actually paint the character. The problem with the current code is that the width we have been allowing for each grapheme is NOT ENOUGH for some of them - especially when we pull in a ***substituted*** glyph from the colour emoji font as the widths of glyphs in THAT font is not necessarily the same as those of the font that we took the `QFontMetrics::averageCharWidth()` from.